pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists)
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight += Self::on_author_noted(orchestrator_chain_author);
+            }
+
+            // Update inactive collator records only after a session has ended
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
+                total_weight +=
+                    Self::process_ended_session(current_unprocessed_session, current_session);
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn process_ended_session(
+            unprocessed_session_id: SessionIndex,
+            active_session_id: SessionIndex,
+        ) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 3);
+            ActiveCollators::<T>::insert(
+                unprocessed_session_id,
+                <ActiveCollatorsForCurrentSession<T>>::get(),
+            );
+
+            <ActiveCollatorsForCurrentSession<T>>::put(BoundedVec::new());
+
+            // Cleanup active collator info for sessions that are older than the maximum allowed
+            let minimum_sessions_required = T::MaxInactiveSessions::get() + 1;
+            if active_session_id >= minimum_sessions_required {
+                total_weight += T::DbWeight::get().writes(1);
+                let _ = <crate::pallet::ActiveCollators<T>>::remove(
+                    active_session_id - minimum_sessions_required,
+                );
+            }
+            <LastUnprocessedSession<T>>::put(active_session_id);
+            total_weight
+        }
+        pub fn on_author_noted(author: T::CollatorId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    if !active_collators.contains(&author) {
+                        total_weight += T::DbWeight::get().writes(1);
+                        active_collators
+                            .try_push(author)
+                            .map_err(|_| Error::<T>::MaxCollatorsPerSessionReached)?;
+                    }
+                    Ok(())
+                },
+            );
+            total_weight
+        }
+    }
+}
+
+impl<T: Config> NodeActivityTrackingHelper<T::CollatorId> for Pallet<T> {
+    fn is_node_inactive(node: &T::CollatorId) -> bool {
+        if !<EnableInactivityTracking<T>>::get() {
```

Comment:
comment why

Reply:
Comment added in [378edb3](https://github.com/moondance-labs/tanssi/pull/922/commits/378edb3d7c2b0c76e2de4991de338a4beb899ebe)